mmdbtype/types.go

```diff
@@ -939,63 +953,96 @@ func writeCtrlByte(w writer, t DataType) (int64, error) {
+func decodeDataTypeValue(decoder *mmdbdata.Decoder, cache map[uint]DataType) (DataType, error) {
 	kind, err := decoder.PeekKind()
 	if err != nil {
 		return nil, fmt.Errorf("peeking kind: %w", err)
 	}
 
+	// Only check cache if provided and the type is worth caching
+	// This avoids unnecessary map lookups for scalar types in tight loops
+	var offset uint
+	if cache != nil && isCacheableKind(kind) {
+		offset = decoder.Offset()
+		if cached, ok := cache[offset]; ok {
+			return cached, nil
+		}
+	}
+
+	var value DataType
 	switch kind {
 	case mmdbdata.KindString:
-		var value String
-		err := value.UnmarshalMaxMindDB(decoder)
-		return value, err
+		var v String
+		err = v.UnmarshalMaxMindDB(decoder)
+		value = v
 	case mmdbdata.KindFloat64:
-		var value Float64
-		err := value.UnmarshalMaxMindDB(decoder)
-		return value, err
+		var v Float64
+		err = v.UnmarshalMaxMindDB(decoder)
+		value = v
 	case mmdbdata.KindBytes:
-		var value Bytes
-		err := value.UnmarshalMaxMindDB(decoder)
-		return value, err
+		var v Bytes
+		err = v.UnmarshalMaxMindDB(decoder)
+		value = v
 	case mmdbdata.KindUint16:
-		var value Uint16
-		err := value.UnmarshalMaxMindDB(decoder)
-		return value, err
+		var v Uint16
+		err = v.UnmarshalMaxMindDB(decoder)
+		value = v
 	case mmdbdata.KindUint32:
-		var value Uint32
-		err := value.UnmarshalMaxMindDB(decoder)
-		return value, err
+		var v Uint32
+		err = v.UnmarshalMaxMindDB(decoder)
+		value = v
 	case mmdbdata.KindInt32:
-		var value Int32
-		err := value.UnmarshalMaxMindDB(decoder)
-		return value, err
+		var v Int32
+		err = v.UnmarshalMaxMindDB(decoder)
+		value = v
 	case mmdbdata.KindUint64:
-		var value Uint64
-		err := value.UnmarshalMaxMindDB(decoder)
-		return value, err
+		var v Uint64
+		err = v.UnmarshalMaxMindDB(decoder)
+		value = v
 	case mmdbdata.KindUint128:
-		var value Uint128
-		err := value.UnmarshalMaxMindDB(decoder)
-		return &value, err // Return pointer for Uint128
+		var v Uint128
+		err = v.UnmarshalMaxMindDB(decoder)
+		value = &v // Return pointer for Uint128
 	case mmdbdata.KindBool:
-		var value Bool
-		err := value.UnmarshalMaxMindDB(decoder)
-		return value, err
+		var v Bool
+		err = v.UnmarshalMaxMindDB(decoder)
+		value = v
 	case mmdbdata.KindFloat32:
-		var value Float32
-		err := value.UnmarshalMaxMindDB(decoder)
-		return value, err
+		var v Float32
+		err = v.UnmarshalMaxMindDB(decoder)
+		value = v
 	case mmdbdata.KindMap:
-		var value Map
-		err := value.UnmarshalMaxMindDB(decoder)
-		return value, err
+		var v Map
+		err = v.unmarshalMaxMindDB(decoder, cache)
+		value = v
 	case mmdbdata.KindSlice:
-		var value Slice
-		err := value.UnmarshalMaxMindDB(decoder)
-		return value, err
+		var v Slice
+		err = v.unmarshalMaxMindDB(decoder, cache)
+		value = v
 	default:
 		return nil, fmt.Errorf("unsupported data type: %v", kind)
 	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Store in cache if provided the offset was captured earlier
+	// when we checked the cache
```

Comment:
Could you clarify this comment? E.g.
> Store in cache if provided _and_ the offset was captured earlier ...

Also, is it correct to say that `decoder.Offset()` above would't normally return `0` ?

Reply:
Re: 0, I forgot that the offsets  are from the start of the data section for this reader; it varies between the different implementations. Using `0` like this could mean that we don't cache that first value, which probably doesn't matter too much overall, but it is easy enough to fix.